miqcli/api.py

```diff
@@ -28,125 +29,194 @@
 class ClientAPI(object):
     """ManageIQ API client class.
 
-    Class interfaces with ManageIQ server. The underlying component performing
-    the interactions with the server is the ManageIQ Python API Client.
+    Class interfaces with ManageIQ server. The underlying component
+    performing the interactions with the server is the ManageIQ Python
+    API Client.
 
     https://github.com/ManageIQ/manageiq-api-client-python
 
-    This library is used to establish connection to the server defined, talk
-    with all the servers collections and perform CRUD operations.
+    This library is used to establish connection to the server defined,
+    talk with all the servers collections and perform CRUD operations.
 
-    Each CLI command will reference this class to perform its collections
-    actions.
+    Each CLI command will reference this class to perform its
+    collections actions.
     """
 
     def __init__(self, settings):
         """Constructor."""
-        self._settings = settings
+
+        # create miqcli folder if it doesn't exist
+        try:
+            os.makedirs(os.path.dirname(AUTHDIR))
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise RuntimeError('Error creating user miqcli folder.')
+
+        # url details
+        self._url = settings.get('url', None)
+        if self._url is None:
+            self._url = 'https://localhost:8443'
```

Comment:
Shouldn't this url be taken from DEFAULT_COFNIG constants

Reply:
sorry looking at the code closer, the DEFAULT_CONFIG is already used in main.py, so values will already be set by default, so you don't need the check for None or having the get set the default value to None if not set, because url should already be set (same for enable_ssl_verfiy, username, and password, no need to set the default value to None, just get the value.

Reply:
With regards to the checking `None`, what if the client API is used out of the click context, let's say, someone importing the class in another app?

Reply:
yeah that makes sense, so I think the only change needed is probably getting default values using the DEFAULT_CONFIG from constants file (url, user, password, enable_ssl_verify)

Reply:
If we add default with DEFAULT_CONFIG['username'] from the constants, and for some reason the constant changes to `user` for example, we break the api module. For that reason we have to add a KeyError exception in the constructor of the client API. Is that what you mean? To use DEFAULT_CONFIG['url']? And use that for the other settings - user, password, etc?

Reply:
I think we just need to set the default values to follow the cli, so the cli and api behave the same, all the current checks currently in there are fine (if using api only, they should be passing the settings w/correct values), but if not set the defaults should be the same for the cli and api.  So I think we remove the static reference for 'https://localhost:8443' and use the DEFAULT_CONFIG_URL and same for the username/password/enable_ssl_verify, if they don't set these values use our set default values (self._verify_ssl = settings.get('enable_ssl_verify', DEFAULT_CONFIG['enable_ssl_verify'])).